### PR TITLE
Define domain custom errors

### PR DESCRIPTION
**Pull request by paulobressan**

### CODE REVIEW

File: src/domain/project/command.rs [hunk 1]
```diff
@@ -159,7 +159,7 @@ pub async fn verify_secret(
 fn assert_credential(credential: &Credential) -> Result<UserId> {
     match credential {
         Credential::Auth0(user_id) => Ok(user_id.into()),
-        Credential::ApiKey(_) => bail!("rpc doesnt support api-key"),
+        Credential::ApiKey(_) => Err(Error::Unauthorized),
```
Comment (gonzalezzfelipe) [inline src/domain/project/command.rs:162]: I think this one should be `Unauthenticated`

File: src/domain/project/command.rs [hunk 2]
```diff
@@ -170,10 +170,15 @@ async fn assert_permission(
     match credential {
         Credential::Auth0(user_id) => {
             let result = cache.find_user_permission(user_id, project_id).await?;
-            ensure!(result.is_some(), "user doesnt have permission");
+            if result.is_none() {
+                return Err(Error::PermissionDenied(
```
Comment (gonzalezzfelipe) [inline src/domain/project/command.rs:174]: This one should be `Unauthorized`

File: src/domain/project/command.rs [hunk 3]
```diff
@@ -170,10 +170,15 @@ async fn assert_permission(
...
+                    "user doesnt have permission".into(),
+                ));
+            }
+
             Ok(())
         }
-        Credential::ApiKey(_) => bail!("rpc doesnt support api-key"),
+        Credential::ApiKey(_) => Err(Error::PermissionDenied("rpc doesnt support api-key".into())),
```
Comment (gonzalezzfelipe) [inline src/domain/project/command.rs:181]: Here `Unauthenticated` as well.

**Review by gonzalezzfelipe — COMMENTED** [review]
Just as a general comment: Unauthenticated means "I don't know who you are" and Unauthorized means "I know who you are, but you are not allowed to do this". I think in this PR you are using `Unauthorized` when you actually mean `Unauthenticated` and `PermissionDenied` should be `Unauthorized`

File: src/domain/project/command.rs (hunk 2 above)
Comment (paulobressan) [reply src/domain/project/command.rs]: Changed, tks feli

File: src/domain/project/command.rs (hunk 3 above)
Comment (paulobressan) [reply src/domain/project/command.rs]: The credential is already valid and this assert will only guarantee that only Auth0 can use the project functions.

**Pull request merged**